### PR TITLE
nixos/unl0kr: add settings to the unl0kr module

### PR DESCRIPTION
**Pull request by hustlerone**
## Description of changes

This PR centers around the service aspect of unl0kr in NixOS (despite the branch name suggesting otherwise). While it would've also been great to also [bump the version](https://github.com/NixOS/nixpkgs/pull/319126#issuecomment-2331351623) I feel like this is a much easier and important contribution, since the most notable changes in between v2.0.0 and v3.2.0 are slight changes in configuration attributes, a new theme and a (presumably hacky but operational) bugfix

Credit to https://github.com/Electrostasy for most of the heavy lifting.

This PR also removes tomfitzhenry as the maintainer of unl0kr at his request.

## Things done

- Built on platform(s)
  - [x] x86_64-linux
  - [ ] aarch64-linux
  - [ ] x86_64-darwin
  - [ ] aarch64-darwin
- For non-Linux: Is sandboxing enabled in `nix.conf`? (See [Nix manual](https://nixos.org/manual/nix/stable/command-ref/conf-file.html))
  - [ ] `sandbox = relaxed`
  - [ ] `sandbox = true`
- [ ] Tested, as applicable:
  - [NixOS test(s)](https://nixos.org/manual/nixos/unstable/index.html#sec-nixos-tests) (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
  - and/or [package tests](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#package-tests)
  - or, for functions and "core" functionality, tests in [lib/tests](https://github.com/NixOS/nixpkgs/blob/master/lib/tests) or [pkgs/test](https://github.com/NixOS/nixpkgs/blob/master/pkgs/test)
  - made sure NixOS tests are [linked](https://github.com/NixOS/nixpkgs/blob/master/pkgs/README.md#linking-nixos-module-tests-to-a-package) to the relevant packages
- [ ] Tested compilation of all packages that depend on this change using `nix-shell -p nixpkgs-review --run "nixpkgs-review rev HEAD"`. Note: all changes have to be committed, also see [nixpkgs-review usage](https://github.com/Mic92/nixpkgs-review#usage)
- [x] Tested basic functionality of all binary files (usually in `./result/bin/`)
- [24.11 Release Notes](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2411.section.md) (or backporting [23.11](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2311.section.md) and [24.05](https://github.com/NixOS/nixpkgs/blob/master/nixos/doc/manual/release-notes/rl-2405.section.md) Release notes)
  - [ ] (Package updates) Added a release notes entry if the change is major or breaking
  - [ ] (Module updates) Added a release notes entry if the change is significant
  - [ ] (Module addition) Added a release notes entry if adding a new NixOS module
- [x] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/CONTRIBUTING.md).

---

Add a :+1: [reaction] to [pull requests you find important].

[reaction]: https://github.blog/2016-03-10-add-reactions-to-pull-requests-issues-and-comments/
[pull requests you find important]: https://github.com/NixOS/nixpkgs/pulls?q=is%3Aopen+sort%3Areactions-%2B1-desc

### CODE REVIEW

File: nixos/modules/system/boot/unl0kr.nix [hunk 1]
```diff
@@ -10,10 +16,32 @@ in
...
+        			}
+            	'';
+      default = { };
+      type = lib.types.submodule { freeformType = settingsFormat.type; };
+    };
   };
 
   config = lib.mkIf cfg.enable {
-    meta.maintainers = with lib.maintainers; [ tomfitzhenry ];
+    meta.maintainers = with lib.maintainers; [ hustlerone ];
```
[the remark this answers is not in the document]
Comment (hustlerone) [reply nixos/modules/system/boot/unl0kr.nix:44]: https://github.com/NixOS/nixpkgs/pull/333616

File: pkgs/by-name/un/unl0kr/package.nix [hunk 2]
```diff
@@ -52,7 +52,7 @@ stdenv.mkDerivation (finalAttrs: {
     mainProgram = "unl0kr";
     homepage = "https://gitlab.com/cherrypicker/unl0kr";
     license = licenses.gpl3Plus;
-    maintainers = with maintainers; [ tomfitzhenry ];
+    maintainers = with maintainers; [ hustlerone ];
```
[the remark this answers is not in the document]
Comment (hustlerone) [reply pkgs/by-name/un/unl0kr/package.nix:55]: https://github.com/NixOS/nixpkgs/pull/333616

**Comment by hustlerone** [comment]
successfully squashed

File: nixos/modules/system/boot/unl0kr.nix [hunk 3]
```diff
@@ -10,10 +16,32 @@ in
...
+        See `unl0kr.conf(5)` for supported values.
+
+        Alternatively, visit `https://gitlab.com/postmarketOS/buffybox/-/blob/unl0kr-2.0.0/unl0kr.conf`
+      '';
+
+      example = lib.literalExpression ''
+        {
+            general.animations = true;
+            theme = {
+            default = "adwaita-dark";
```
Comment (Electrostasy) [inline nixos/modules/system/boot/unl0kr.nix:33]: Nitpick, but the indentation seems to be off on this line.

Comment (hustlerone) [reply nixos/modules/system/boot/unl0kr.nix:33]: that's odd this is outdated

Comment (hustlerone) [reply nixos/modules/system/boot/unl0kr.nix]: resolved, with the appropriate settings for that version

**Comment by Frontear** [comment]
#333616
#319126

Notifying since this PR seems to try and handle these other two as well. Can we get discussion from the respective PR openers, including potential closure of the older PRs if this one suffices?

**Comment by hustlerone** [comment]
https://github.com/NixOS/nixpkgs/pull/333616 has been accepted and I'll need to rebase and fix the conflicts

**Comment by hustlerone** [comment]
> #333616 #319126
> 
> Notifying since this PR seems to try and handle these other two as well. Can we get discussion from the respective PR openers, including potential closure of the older PRs if this one suffices?

This PR only handles the service side of things and version bumps can be done separately

**Comment by hustlerone** [comment]
Don't merge this yet, I have to make sure certain modules are loaded for the correct operation of unl0kr

**Comment by hustlerone** [comment]
any drawbacks of boot.initrd.availableKernelModules compared to boot.initrd.kernelModules?

**Comment by Electrostasy** [comment]
> Don't merge this yet, I have to make sure certain modules are loaded for the correct operation of unl0kr

I noticed I had to add evdev for unl0kr in initrd (iirc it is listed as a dependency in the project's readme), but also hid_multitouch, i2c_hid_acpi and intel_lpss_pci for touchscreen support.

On my device, touchscreen is apparently connected over i2c, and that was the bare minimum I needed to get touch working (both drivers and actually powering up the i2c devices). I believe these are rather dependent on how touchscreen is wired up on a particular device, and it may be different for other hardware. I found https://www.kernelconfig.io/index.html very useful for figuring out what is loaded and why when I set my unl0kr up, though.

> any drawbacks of boot.initrd.availableKernelModules compared to boot.initrd.kernelModules?

As far as I understand, `boot.initrd.availableKernelModules` should just add the modules to the initrd image, but `boot.initrd.kernelModules` will add them *and* load them on startup. In practice, they are functionally the same, as modules will be loaded if they are available and required in my experience.

**Comment by hustlerone** [comment]
> > Don't merge this yet, I have to make sure certain modules are loaded for the correct operation of unl0kr
> 
> I noticed I had to add evdev for unl0kr in initrd (iirc it is listed as a dependency in the project's readme), but also hid_multitouch, i2c_hid_acpi and intel_lpss_pci for touchscreen support.
> 
> On my device, touchscreen is apparently connected over i2c, and that was the bare minimum I needed to get touch working (both drivers and actually powering up the i2c devices). I believe these are rather dependent on how touchscreen is wired up on a particular device, and it may be different for other hardware. I found https://www.kernelconfig.io/index.html very useful for figuring out what is loaded and why when I set my unl0kr up, though.
> 
> > any drawbacks of boot.initrd.availableKernelModules compared to boot.initrd.kernelModules?
> 
> As far as I understand, `boot.initrd.availableKernelModules` should just add the modules to the initrd image, but `boot.initrd.kernelModules` will add them _and_ load them on startup. In practice, they are functionally the same, as modules will be loaded if they are available and required in my experience.

My concern is that there could be a complaint about bloat because maybe only one of them is needed

**Comment by hustlerone** [comment]
I'm going to test on my machine right now

**Comment by hustlerone** [comment]
Tested, seems to work on the Steam Deck

**Comment by Electrostasy** [comment]
> My concern is that there could be a complaint about bloat because maybe only one of them is needed

Yes, due to potentially different hardware configurations and considering that the initrd image size will increase, it would be reasonable to only require evdev to be added by the NixOS module IMO. Maybe also the generic touch modules?

I haven't seen the i2c-designware modules before, so I will likely test this myself with them tomorrow.

**Comment by hustlerone** [comment]
> > My concern is that there could be a complaint about bloat because maybe only one of them is needed
> 
> Yes, due to potentially different hardware configurations and considering that the initrd image size will increase, it would be reasonable to only require evdev to be added by the NixOS module IMO. Maybe also the generic touch modules?
> 
> I haven't seen the i2c-designware modules before, so I will likely test this myself with them tomorrow.

I was more concerned about the RAM usage. If anyone wants to strip down the list they're free to use `lib.lists.remove` when they define boot.initrd.kernelModules

**Comment by hustlerone** [comment]
Oh yeah, and the Steam Deck requires 5 out of the 8 modules I've defined

**Comment by hustlerone** [comment]
This PR is ready to be merged now

**Comment by ElvishJerricco** [comment]
> If anyone wants to strip down the list they're free to use lib.lists.remove when they define boot.initrd.kernelModules

not really. The module system doesn't have a way to *remove* definitions from another module, other than using `mkForce` to override the value completely (which means copying any other definitions that you would have needed to keep).

File: nixos/modules/system/boot/unl0kr.nix [hunk 4]
```diff
@@ -10,18 +16,58 @@ in
...
+    boot.initrd.availableKernelModules = lib.optionals cfg.enable [
+      "hid-multitouch"
+      "hid-generic"
+      "usbhid"
+
+      "i2c-designware-core"
+      "i2c-designware-platform"
+      "i2c-hid-acpi"
+
+      "usbtouchscreen"
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/unl0kr.nix:65]: What is the rationale behind these ones specifically? `evdev` seems required, sure. And I suppose `hid-multitouch` and `hid-generic` make pretty obvious sense.

It's not that I'm opposed to adding the others. It just seems like it might be favoring a particular configuration? Feels like we should either be providing a minimal set of modules or a pretty comprehensive set.

Comment (ElvishJerricco) [reply nixos/modules/system/boot/unl0kr.nix:65]: Though, the change in the size of the initrd is negligible so I think we can just do it like this and forget about it.

File: maintainers/maintainer-list.nix [hunk 5]
```diff
@@ -3200,9 +3200,7 @@
     github = "cageyv";
     githubId = 51059484;
     name = "Vladmir Samoylov";
-    keys = [
-      { fingerprint = "8916 F727 734E 77AB 437F  A33A 19AB 76F5 CEE1 1392"; }
-    ];
+    keys = [ { fingerprint = "8916 F727 734E 77AB 437F  A33A 19AB 76F5 CEE1 1392"; } ];
```
Comment (Aleksanaa) [inline maintainers/maintainer-list.nix:3203]: Please don't touch unrelated lines

**Comment by Electrostasy** [comment]
> I haven't seen the i2c-designware modules before, so I will likely test this myself with them tomorrow.

unl0kr does load with this PR on my 2 in 1 laptop, however, I do not have any touchscreen support unless I additionally include `intel_lpss_pci` in `boot.initrd.availableKernelModules`.

I can add it to my specific configuration (since it is specific to my hardware), but I still have some doubts about including kernel modules in the initrd only for certain hardware here. I suppose it would be better to add them to get as much compatibility as we feasibly can, though.

File: nixos/modules/system/boot/unl0kr.nix (hunk 4 above)
Comment (hustlerone) [reply nixos/modules/system/boot/unl0kr.nix:65]: I'm adding anything that might be related to touch support (The comments in Jovian NixOS suggest that the i2c modules are for touchscreen support on Steam Deck)

File: maintainers/maintainer-list.nix (hunk 5 above)
Comment (hustlerone) [reply maintainers/maintainer-list.nix:3203]: This is the work of nixfmt and anyone that adds themselves to the contributors list may end up in this same situation if I end up undoing this change

**Comment by hustlerone** [comment]
> > I haven't seen the i2c-designware modules before, so I will likely test this myself with them tomorrow.
> 
> unl0kr does load with this PR on my 2 in 1 laptop, however, I do not have any touchscreen support unless I additionally include `intel_lpss_pci` in `boot.initrd.availableKernelModules`.
> 
> I can add it to my specific configuration (since it is specific to my hardware), but I still have some doubts about including kernel modules in the initrd only for certain hardware here. I suppose it would be better to add them to get as much compatibility as we feasibly can, though.

That is my intention. After all, this module is supposed to have unl0kr painlessly work on as much hardware as possible.

File: nixos/modules/system/boot/unl0kr.nix (hunk 4 above)
Comment (hustlerone) [reply nixos/modules/system/boot/unl0kr.nix:65]: My intent is to support as much hardware as possible so this module is as painless as possible

**Comment by hustlerone** [comment]
@Aleksanaa I really have to use nixfmt so editorconfig does not complain.

**Comment by hustlerone** [comment]
more drivers added

**Comment by hustlerone** [comment]
Tested and confirmed working on wacom

**Comment by ElvishJerricco** [comment]
Ok, the `intel*` is sort of a bridge too far for me, since it's explicitly vendor-specific. I don't mind it being enabled by default but there needs to be a way to disable it.

Here's my solution. We should have an option called something like `enableOptionalModules` which defaults to `true`. Now, this shouldn't interfere with the inclusion of `evdev`. It seems to me that `evdev` is a hard requirement and the lack of it is a current bug. So we probably have to do some `mkMerge` magic to accomplish this. But I think having an option enabled by default that is easy to disable would be important here

**Comment by hustlerone** [comment]
> Ok, the `intel*` is sort of a bridge too far for me, since it's explicitly vendor-specific. I don't mind it being enabled by default but there needs to be a way to disable it.
> 
> Here's my solution. We should have an option called something like `enableOptionalModules` which defaults to `true`. Now, this shouldn't interfere with the inclusion of `evdev`. It seems to me that `evdev` is a hard requirement and the lack of it is a current bug. So we probably have to do some `mkMerge` magic to accomplish this. But I think having an option enabled by default that is easy to disable would be important here

Where do we draw the line?

**Comment by hustlerone** [comment]
(I'm saying this because wacom is also vendor-specific)

File: nixos/modules/system/boot/unl0kr.nix [hunk 6]
```diff
@@ -10,18 +16,65 @@ in
...
         message = "boot.initrd.unl0kr is only supported with boot.initrd.systemd.";
       }
+      {
+        assertion = !config.boot.plymouth.enable;
+        message = "unl0kr will not work properly if plymouth is enabled.";
+      }
+      {
+        assertion = !config.hardware.amdgpu.initrd.enable;
+        message = "unl0kr may not work properly if video drivers are loaded on stage 1.";
+      }
```
Comment (ElvishJerricco) [inline nixos/modules/system/boot/unl0kr.nix:57]: This doesn't seem like something we should be asserting. It's going to depend on the specific hardware, isn't it?

**Comment by ElvishJerricco** [comment]
> (I'm saying this because wacom is also vendor-specific)

@hustlerone Yea, i would say that anything specific to a particular brand or vendor would be behind such a `enableOptionalModules` option.

File: nixos/modules/system/boot/unl0kr.nix
Comment (JohnRTitor) [inline nixos/modules/system/boot/unl0kr.nix:1]: Commit message should be updated.
```
nixos/unl0kr: add settings option ........
```

File: nixos/modules/system/boot/unl0kr.nix (hunk 6 above)
Comment (hustlerone) [reply nixos/modules/system/boot/unl0kr.nix:57]: it does both on Steam Deck and desktop machine

**Comment by hustlerone** [comment]
I'm going to do the requested changes, stand by

**Comment by ElvishJerricco** [comment]
@ofborg test systemd-initrd-luks-unl0kr
